q2_fmt/_peds.py

```diff
@@ -0,0 +1,153 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2022-2022, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import qiime2
+import pandas as pd
+
+
+def sample_peds(table: pd.DataFrame, metadata: qiime2.Metadata,
+                time_column: str, reference_column: str, subject_column: str,
+                filter_missing_references: bool = False,
+                drop_incomplete_subjects: bool = False) -> (pd.DataFrame):
+    # TODO: Make incomplete samples possible move this to heatmap
+    metadata = metadata.to_dataframe()
+
+    try:
+        num_timepoints = metadata[time_column].dropna().unique().size
+    except Exception as e:
+        raise KeyError('There was an error finding %s in the metadata'
+                       % time_column) from e
+    try:
+        metadata[time_column].dropna().astype(int)
+    except Exception as e:
+        raise KeyError('%s must be numeric' % time_column) from e
+    try:
+        subject_series = metadata[subject_column]
+    except Exception as e:
+        raise KeyError('There was an error finding %s in the metadata'
+                       % subject_column) from e
+    used_subject_series = subject_series[~metadata[time_column].isna()]
+    subject_occurrence_df = (used_subject_series.value_counts()
+                             .to_frame())
+    if (subject_occurrence_df[subject_column] != num_timepoints).any():
+        if drop_incomplete_subjects:
+            subject_to_keep = (subject_occurrence_df
+                               .loc[subject_occurrence_df[subject_column]
+                                    == num_timepoints].index)
+            metadata = metadata[metadata[subject_column].isin(subject_to_keep)]
+        elif (subject_occurrence_df[subject_column] < num_timepoints).any():
+            incomplete_subjects = (subject_occurrence_df
+                                   .loc[subject_occurrence_df[subject_column]
+                                        < num_timepoints].index).to_list()
+            raise KeyError('Missing timepoints for associated subjects.'
+                           ' Please make sure that all subjects have all'
+                           ' timepoints or use drop_incomplete_subjects'
+                           ' parameter. The incomplete subjects were %s'
+                           % incomplete_subjects)
+        elif (subject_occurrence_df[subject_column] > num_timepoints).any():
```

Comment:
This idea is better caught by ensuring subjects are unique within timepoints. As the only way to hit this branch would be if the data was like the above table, but with extra repeated measures for a subject in a timepoint, which means that this error message isn't really the core of the issue.

Otherwise, if subjects were unique within a timepoint, I think it would be impossible for a subject to have more observations than the number of timepoints, as they are defined by those samples which have timepoints.